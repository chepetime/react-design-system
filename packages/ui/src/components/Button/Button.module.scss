@use "./../../styles/colors.scss";
@use "./../../styles/mixins";

:root {
  --button-background-color: #4cbaea;
  --button-background-color-hover: #3888ab;
  --button-disabled-background-color: #cccccc;
}

:global :root .dark {
  --button-background-color: #ea4c89;
  --button-background-color-hover: #a3345f;
  --button-disabled-background-color: #cccccc;
}

.Button {
  all: revert;
  -webkit-appearance: none;

  font-family: "Helvetica Neue", Helvetica, Arial, sans-serif, sans-serif;
  font-size: 1rem;
  font-weight: 400;
  font-style: normal;

  width: 100%;
  min-width: 5rem;
  min-height: 2.5rem;
  margin: 0;
  padding: 0 1rem;

  cursor: pointer;
  -webkit-user-select: none;
  user-select: none;
  transition: color 100ms;
  text-align: center;
  vertical-align: baseline;
  text-decoration: none;

  text-align: center;
  white-space: nowrap;

  color: inherit;
  color: #fff;
  border: none;
  border-radius: 0.5rem;
  background: var(--button-background-color);
  box-shadow: none;

  transition: all ease-in-out 0.3s;

  @include mixins.respond-to("md") {
    width: auto;
  }

  /* If device has hover capabilities */
  @include mixins.feature-to("hover") {
    &:hover:not(:disabled) {
      background-color: var(--button-background-color-hover);
    }
  }

  /* Focus */
  &:focus:not(:focus-visible) {
    outline: none;
  }

  /* Focus */
  &:focus-visible {
    outline: none;
    box-shadow: hsl(27deg 39% 95%) 0 0 0 2px, hsl(27deg 100% 56%) 0 0 0 4px;
  }

  &:disabled,
  &[disabled] {
    background-color: var(--button-disabled-background-color);
    color: #666666;
    pointer-events: none;
  }
}
