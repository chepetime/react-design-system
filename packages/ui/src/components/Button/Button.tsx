import React from "react";
import classnames from "classnames";
import css from "./Button.module.scss";

export interface ButtonProps {
  children?: React.ReactNode;
}

export const Button = React.forwardRef<HTMLDivElement, ButtonProps>(
  (props: ButtonProps, ref: React.Ref<HTMLDivElement>) => {
    return (
      <div className={classnames(css.Button)} ref={ref}>
        <p>Button component working!</p>
        <div>{props.children}</div>
      </div>
    );
  }
)

Button.displayName = "Button";
