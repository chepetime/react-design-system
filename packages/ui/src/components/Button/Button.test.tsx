import React from "react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom/extend-expect";
import { Button } from "./Button";
import { render, screen } from "@testing-library/react";

describe("Button", () => {
  it("renders without errors", () => {
    render(<Button />);
    // Assert that the component renders without throwing any errors
  });

  it("renders with children", () => {
    render(<Button>Some children</Button>);
    // Assert that the component renders the provided children
    const childrenElement = screen.getByText("Some children");
    expect(childrenElement).toBeInTheDocument();
  });

  it("handles user interactions", () => {
    render(<Button>Some children</Button>);
    // Simulate user interactions with the component
    userEvent.click(screen.getByText("Some children"));
    // Assert the expected outcome of the user interaction
    // For example, check if a certain state has changed or an event handler has been called
  });

  it("applies variant class", () => {
    render(<Button variant="ghost">Ghost Button</Button>);
    const buttonElement = screen.getByText(/Ghost Button/i);
    expect(buttonElement).toHaveClass("Button--ghost");
  });

  it("applies color class", () => {
    render(<Button color="secondary">Secondary Button</Button>);
    const buttonElement = screen.getByText(/Secondary Button/i);
    expect(buttonElement).toHaveClass("Button--secondary");
  });

  it("applies size class", () => {
    render(<Button size="small">Small Button</Button>);
    const buttonElement = screen.getByText(/Small Button/i);
    expect(buttonElement).toHaveClass("Button--small");
  });

  it("handles click event", async () => {
    const handleClick = jest.fn();
    render(<Button onClick={() => handleClick()}>Click Me</Button>);
    const buttonElement = screen.getByText(/Click Me/i);
    await userEvent.click(buttonElement);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it("matches snapshot", () => {
    const { container } = render(<Button />);
    // Assert that the rendered component matches a previously captured snapshot
    expect(container).toMatchSnapshot();
  });
});
