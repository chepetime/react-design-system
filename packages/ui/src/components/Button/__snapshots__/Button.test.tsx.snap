// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button matches snapshot 1`] = `
<div>
  <div
    class="Button"
  >
    <p>
      Button component working!
    </p>
    <div />
  </div>
</div>
`;
